### PR TITLE
ADXT-187 Implement Test Washer 

### PR DESCRIPTION
**Pull request by KevinFairise2**
### What does this PR do?

Add a TestWasher that can be used in the `test` invoke task.

See: https://docs.google.com/document/d/1DQyuD56Ag2vefQoPObnKR0pkAkiHb54xD590vdmqIaI/edit#heading=h.s2qtmwue4n26
For more details about the TestWasher current behavior

TLDR: Tests that are known to be flaky, either because they are marked using `flaky.Mark` or because they are listed in `flakes.yaml` will be ignored. If only flaky tests fail in a test job, the job will be reported as successfull.
CI Visibility always get the failures reported

### CODE REVIEW

File: tasks/github_tasks.py [hunk 1]
```diff
@@ -68,6 +68,7 @@ def trigger_macos(
     retry_download=3,
     retry_interval=10,
     fast_tests=None,
+    test_washer=None,
```
Comment (FlorentClarret) [inline tasks/github_tasks.py:71]: nit: you used `False` as the default value in the other task. We might change it to be consistent

File: tasks/go_test.py [hunk 2]
```diff
@@ -296,6 +298,34 @@ def sanitize_env_vars():
...
+        return True
+
+    if test_washer:
+        tw = TestWasher()
+        should_succeed = tw.process_module_results(test_results)
+        if should_succeed:
+            print(
+                color_message("All failing tests are known to be flaky, marking the test job as successful", "orange")
+            )
+            return True
```
Comment (FlorentClarret) [inline tasks/go_test.py:323]: You do not have any return if `should_succeed` is `False`. I think you can simply get rid of the `else` statement

File: tasks/go_test.py [hunk 3]
```diff
@@ -296,6 +298,34 @@ def sanitize_env_vars():
             del os.environ[env]
 
 
+def process_test_result(test_results, junit_tar, flavor, test_washer):
```
Comment (FlorentClarret) [inline tasks/go_test.py:301]: Could we add the types for the parameters?

File: tasks/testwasher.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+class TestWasher:
+    test_output_json_file = "module_test_output.json"
+    flaky_test_indicator = "flakytest: this is a known flaky test"
+    flakes_file_path = "flakes.yaml"
+
+    known_flaky_tests = defaultdict(set)
+
+    def __init__(self, test_output_json_file=None, flaky_test_indicator=None, flakes_file_path=None):
+        self.test_output_json_file = test_output_json_file if test_output_json_file else "module_test_output.json"
```
Comment (FlorentClarret) [inline tasks/testwasher.py:16]: You could set `"module_test_output.json"` as the default value for the parameter directly

File: tasks/testwasher.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        self.flaky_test_indicator = (
+            flaky_test_indicator if flaky_test_indicator else "flakytest: this is a known flaky test"
+        )
+        self.flakes_file_path = flakes_file_path if flakes_file_path else "flakes.yaml"
```
Comment (FlorentClarret) [inline tasks/testwasher.py:20]: Same here in fact

**Review by FlorentClarret — COMMENTED** [review]
Looks nice! I left some comments and I'll have a deeper look after I manage to catch up with all the stuff I missed during my PTOs :)

File: tasks/go_test.py (hunk 2 above)
Comment (KevinFairise2) [reply tasks/go_test.py:323]: Nice catch!

**Comment by pducolin** [comment]
I think this branch missed alignment with [kserrania/refactor-test-task](https://github.com/DataDog/datadog-agent/pull/25015/commits/d45dfcebbe79a70009346c25e61162bf92554a63) that was merged in the mean time, would it be possible to reset the branch on main and cherry pick your changes?

File: tasks/testwasher.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    def parse_flaky_file(self):
+        """
+        Parse the flakes.yaml file and add the tests listed there to the kown flaky tests list
+        """
+        with open(self.flakes_file_path) as f:
+            flakes = yaml.safe_load(f)
+            if not flakes:
+                return
+            for package, tests in flakes.items():
+                self.known_flaky_tests[f"github.com/DataDog/datadog-agent/{package}"] = set(tests)
```
Comment (pducolin) [inline tasks/testwasher.py:64]: **❓ question**
Should we merge with existing tests instead? 
```suggestion
			    if f"github.com/DataDog/datadog-agent/{package}" in self.known_flaky_tests:
			    		tests = self.known_flaky_tests[f"github.com/DataDog/datadog-agent/{package}"].union(tests)
                self.known_flaky_tests[f"github.com/DataDog/datadog-agent/{package}"] = set(tests)
```

**Review by pducolin — APPROVED** [review]
Very nice!

File: tasks/testwasher.py (hunk 6 above)
Comment (KevinFairise2) [reply tasks/testwasher.py:64]: Ah yes, I can do that. I expected the `flakes.yaml` to contain only one key for each package. But that's safer to not assume that and to merge the sets!

**Comment by KevinFairise2** [comment]
> I think this branch missed alignment with [kserrania/refactor-test-task](https://github.com/DataDog/datadog-agent/pull/25015/commits/d45dfcebbe79a70009346c25e61162bf92554a63) that was merged in the mean time, would it be possible to reset the branch on main and cherry pick your changes?

I pulled Kylian's branch on mine because it was not merged yet when I started to work on this. It makes the commit history pretty bad. I will try to improve it

**Comment by KevinFairise2** [comment]
Just force pushed the branch to improve the commit history

File: tasks/go_test.py [hunk 7]
```diff
@@ -296,6 +298,33 @@ def sanitize_env_vars():
             del os.environ[env]
 
 
+def process_test_result(test_results: ModuleTestResult, junit_tar: str, flavor: AgentFlavor, test_washer: bool):
```
Comment (CelianR) [inline tasks/go_test.py:301]: **🥜 nitpick:** ‏
```suggestion
def process_test_result(test_results: ModuleTestResult, junit_tar: str, flavor: AgentFlavor, test_washer: bool) -> bool:
```

File: tasks/go_test.py [hunk 8]
```diff
@@ -296,6 +298,33 @@ def sanitize_env_vars():
...
+    if junit_tar:
+        junit_files = []
+        for module_test_result in test_results:
+            if module_test_result.junit_file_path:
+                junit_files.append(module_test_result.junit_file_path)
```
Comment (CelianR) [inline tasks/go_test.py:306]: **🥜 nitpick:** ‏
```suggestion
        junit_files = [module_test_result.junit_file_path for module_test_result in test_results if module_test_result.junit_file_path]
```

File: tasks/testwasher.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    def get_non_flaky_failing_tests(self, module_path):
+        """
+        Parse the test output json file and compute the failing tests and the one known flaky
+        """
+        failing_tests = defaultdict(set)
+        flaky_marked_tests = defaultdict(set)
+        with open(f"{module_path}/{self.test_output_json_file}", encoding='utf-8') as f:
+            for line in f:
+                test_result = json.loads(line)
+                # TODO: tests can be retried we should handle that
```
Comment (CelianR) [inline tasks/testwasher.py:31]: **❓ question:** This will be handled in another PR ?‏

**Review by CelianR — APPROVED** [review]

File: tasks/testwasher.py (hunk 9 above)
Comment (KevinFairise2) [reply tasks/testwasher.py:31]: Good point, I can handle it in this PR. Was a reminder for me to not forget it and I did forget it 😓

**Review by clarkb7 — APPROVED** [review]

**Review by FlorentClarret — APPROVED** [review]
